File: trip-kit/src/test/java/com/skedgo/android/tripkit/TripUpdaterImplTest.java
package com.skedgo.android.tripkit;

import android.content.res.Resources;

import com.google.gson.Gson;
import com.skedgo.android.common.model.RoutingResponse;
import com.skedgo.android.common.model.Trip;
import com.skedgo.android.common.model.TripGroup;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;

import java.util.ArrayList;

import rx.Observable;
import rx.observers.TestSubscriber;

import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class TripUpdaterImplTest {
  @Mock Resources resources;
  @Mock TripUpdateApi api;
  private String apiVersion = "11";
  private TripUpdaterImpl updater;

  @Before public void before() {
    MockitoAnnotations.initMocks(this);
    updater = new TripUpdaterImpl(resources, api, apiVersion, new Gson());
  }

  @Test public void shouldReturnEmpty() {
    final Trip trip = mock(Trip.class);
    final TestSubscriber<Trip> subscriber = new TestSubscriber<>();
    updater.getUpdateAsync(trip).subscribe(subscriber);
    subscriber.awaitTerminalEvent();
    subscriber.assertNoErrors();
    subscriber.assertNoValues();
  }

  @Test public void shouldReturnDisplayTrip() {
    final TripGroup group = mock(TripGroup.class);
    final Trip trip = mock(Trip.class);
    final String updateUrl = "https://goo.gl/f2yv7k";
    when(trip.getUpdateURL()).thenReturn(updateUrl);
    final Trip displayTrip = mock(Trip.class);
    when(group.getDisplayTrip()).thenReturn(displayTrip);
    final RoutingResponse response = mock(RoutingResponse.class);
    when(response.getTripGroupList())
        .thenReturn(new ArrayList<>(singletonList(group)));
    when(api.fetchUpdateAsync(eq(updateUrl), eq(apiVersion)))
        .thenReturn(Observable.just(response));

    final TestSubscriber<Trip> subscriber = new TestSubscriber<>();
    updater.getUpdateAsync(trip).subscribe(subscriber);
    subscriber.awaitTerminalEvent();
    subscriber.assertNoErrors();
    assertThat(subscriber.getOnNextEvents()).containsExactly(displayTrip);
  }

  @Test public void shouldReturnEmptyIfNoGroupsAvailable() {
    final Trip trip = mock(Trip.class);
    final String updateUrl = "https://goo.gl/f2yv7k";
    when(trip.getUpdateURL()).thenReturn(updateUrl);
    final RoutingResponse response = mock(RoutingResponse.class);
    when(response.getTripGroupList()).thenReturn(null);
    when(api.fetchUpdateAsync(eq(updateUrl), eq(apiVersion)))
        .thenReturn(Observable.just(response));

    final TestSubscriber<Trip> subscriber = new TestSubscriber<>();
    updater.getUpdateAsync(trip).subscribe(subscriber);
    subscriber.awaitTerminalEvent();
    subscriber.assertNoErrors();
    subscriber.assertNoValues();
  }

  @Test public void shouldReturnEmptyIfNoTripsAvailable() {
    final Trip trip = mock(Trip.class);
    final String updateUrl = "https://goo.gl/f2yv7k";
    when(trip.getUpdateURL()).thenReturn(updateUrl);
    final TripGroup group = mock(TripGroup.class);
    when(group.getDisplayTrip()).thenReturn(null);
    final RoutingResponse response = mock(RoutingResponse.class);
    when(response.getTripGroupList())
        .thenReturn(new ArrayList<>(singletonList(group)));
    when(api.fetchUpdateAsync(eq(updateUrl), eq(apiVersion)))
        .thenReturn(Observable.just(response));

    final TestSubscriber<Trip> subscriber = new TestSubscriber<>();
    updater.getUpdateAsync(trip).subscribe(subscriber);
    subscriber.awaitTerminalEvent();
    subscriber.assertNoErrors();
    subscriber.assertNoValues();
  }
}

File: TripKitAndroid/src/test/java/com/skedgo/android/tripkit/TripUpdaterImplTest.kt
package com.skedgo.android.tripkit

import android.content.res.Resources
import com.google.gson.Gson
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Java6Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.mock
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import rx.Observable
import rx.observers.TestSubscriber
import skedgo.tripkit.routing.RoutingResponse
import skedgo.tripkit.routing.Trip
import skedgo.tripkit.routing.TripGroup
import java.util.*

@RunWith(RobolectricTestRunner::class)
class TripUpdaterImplTest {
  @Mock
  internal var resources: Resources? = null
  @Mock
  internal var api: TripUpdateApi? = null
  private var updater: TripUpdaterImpl? = null

  @Before fun before() {
    MockitoAnnotations.initMocks(this)
    updater = TripUpdaterImpl(resources!!, api!!, Gson())
  }

  @Test fun shouldReturnEmpty() {
    val trip = mock(Trip::class.java)
    val subscriber = TestSubscriber<Trip>()
    updater!!.getUpdateAsync(trip).subscribe(subscriber)
    subscriber.awaitTerminalEvent()
    subscriber.assertNoErrors()
    subscriber.assertNoValues()
  }

  @Test fun shouldReturnDisplayTrip() {
    val group = mock(TripGroup::class.java)
    val trip = mock(Trip::class.java)
    val updateUrl = "https://goo.gl/f2yv7k"
    whenever(trip.updateURL).thenReturn(updateUrl)
    val displayTrip = mock(Trip::class.java)
    whenever(group.displayTrip).thenReturn(displayTrip)
    val response = mock(RoutingResponse::class.java)
    whenever(response.tripGroupList)
        .thenReturn(ArrayList(listOf(group)))
    whenever(api!!.fetchUpdateAsync(updateUrl))
        .thenReturn(Observable.just(response))

    val subscriber = TestSubscriber<Trip>()
    updater!!.getUpdateAsync(trip).subscribe(subscriber)
    subscriber.awaitTerminalEvent()
    subscriber.assertNoErrors()
    assertThat(subscriber.onNextEvents).containsExactly(displayTrip)
  }

  @Test fun shouldReturnEmptyIfNoGroupsAvailable() {
    val trip = mock(Trip::class.java)
    val updateUrl = "https://goo.gl/f2yv7k"
    whenever(trip.updateURL).thenReturn(updateUrl)
    val response = mock(RoutingResponse::class.java)
    whenever(response.tripGroupList).thenReturn(null)
    whenever(api!!.fetchUpdateAsync(updateUrl))
        .thenReturn(Observable.just(response))

    val subscriber = TestSubscriber<Trip>()
    updater!!.getUpdateAsync(trip).subscribe(subscriber)
    subscriber.awaitTerminalEvent()
    subscriber.assertNoErrors()
    subscriber.assertNoValues()
  }

  @Test fun shouldReturnEmptyIfNoTripsAvailable() {
    val trip = mock(Trip::class.java)
    val updateUrl = "https://goo.gl/f2yv7k"
    whenever(trip.updateURL).thenReturn(updateUrl)
    val group = mock(TripGroup::class.java)
    whenever(group.displayTrip).thenReturn(null)
    val response = mock(RoutingResponse::class.java)
    whenever(response.tripGroupList)
        .thenReturn(ArrayList(listOf(group)))
    whenever(api!!.fetchUpdateAsync(updateUrl))
        .thenReturn(Observable.just(response))

    val subscriber = TestSubscriber<Trip>()
    updater!!.getUpdateAsync(trip).subscribe(subscriber)
    subscriber.awaitTerminalEvent()
    subscriber.assertNoErrors()
    subscriber.assertNoValues()
  }
}